d2layouts/d2sequence/layout.go

```diff
@@ -105,3 +150,129 @@ func addLifelineEdges(g *d2graph.Graph, actors []*d2graph.Object, yStep float64)
+	for _, activation := range activationFromMostNested {
+		// finds the position based on children
+		minChildY := math.Inf(1)
+		maxChildY := math.Inf(-1)
+		for _, child := range activation.ChildrenArray {
+			minChildY = math.Min(minChildY, child.TopLeft.Y)
+			maxChildY = math.Max(maxChildY, child.TopLeft.Y+child.Height)
+		}
+
+		// finds the position if there are edges to this activation box
+		minEdgeY := math.Inf(1)
+		if minRank, exists := sd.minEdgeRank[activation]; exists {
+			minEdgeY = sd.getEdgeY(minRank)
+		}
+		maxEdgeY := math.Inf(-1)
+		if maxRank, exists := sd.maxEdgeRank[activation]; exists {
+			maxEdgeY = sd.getEdgeY(maxRank)
+		}
+
+		// if it is the same as the child top left, add some padding
+		minY := math.Min(minEdgeY, minChildY)
+		if minY == minChildY {
+			minY -= ACTIVATION_BOX_DEPTH_GROW_FACTOR
+		} else {
+			minY -= ACTIVATION_BOX_EDGE_PAD
+		}
+		maxY := math.Max(maxEdgeY, maxChildY)
+		if maxY == maxChildY {
+			maxY += ACTIVATION_BOX_DEPTH_GROW_FACTOR
+		} else {
+			maxY += ACTIVATION_BOX_EDGE_PAD
+		}
+
+		height := math.Max(maxY-minY, DEFAULT_ACTIVATION_BOX_HEIGHT)
+		width := ACTIVATION_BOX_WIDTH + (float64(sd.objectDepth[activation]-1) * ACTIVATION_BOX_DEPTH_GROW_FACTOR)
+		x := rankToX[sd.objectRank[activation]] - (width / 2.)
+		activation.Box = geo.NewBox(geo.NewPoint(x, minY), width, height)
+	}
+}
+
+// routeEdges routes horizontal edges from Src to Dst
+func (sd *sequenceDiagram) routeEdges() {
+	for rank, edge := range sd.edges {
+		isLeftToRight := edge.Src.TopLeft.X < edge.Dst.TopLeft.X
+
+		// finds the proper anchor point based on the edge direction
+		var startX, endX float64
+		if sd.isActor(edge.Src) {
+			startX = edge.Src.Center().X
+		} else if isLeftToRight {
+			startX = edge.Src.TopLeft.X + edge.Src.Width
+		} else {
+			startX = edge.Src.TopLeft.X
+		}
+
+		if sd.isActor(edge.Dst) {
+			endX = edge.Dst.Center().X
+		} else if isLeftToRight {
+			endX = edge.Dst.TopLeft.X
+		} else {
+			endX = edge.Dst.TopLeft.X + edge.Dst.Width
+		}
+
+		if isLeftToRight {
+			startX += ACTIVATION_BOX_EDGE_PAD
+			endX -= ACTIVATION_BOX_EDGE_PAD
+		} else {
+			startX -= ACTIVATION_BOX_EDGE_PAD
+			endX += ACTIVATION_BOX_EDGE_PAD
+		}
+
+		edgeY := sd.getEdgeY(rank)
+		edge.Route = []*geo.Point{
+			geo.NewPoint(startX, edgeY),
+			geo.NewPoint(endX, edgeY),
+		}
+
+		if edge.Attributes.Label.Value != "" {
+			if isLeftToRight {
+				edge.LabelPosition = go2.Pointer(string(label.OutsideTopCenter))
+			} else {
+				// the label will be placed above the edge because the orientation is based on the edge normal vector
+				edge.LabelPosition = go2.Pointer(string(label.OutsideBottomCenter))
+			}
+		}
+	}
+}
+
+func (sd *sequenceDiagram) getEdgeY(rank int) float64 {
+	// +1 so that the first edge has the top padding for its label
+	return ((float64(rank) + 1.) * sd.edgeYStep) + sd.maxActorHeight
+}
+
+func (sd *sequenceDiagram) isActor(obj *d2graph.Object) bool {
+	return obj.Parent == sd.graph.Root
```

Comment:
this isn't right. u can have a very nested sequence diagram.

Reply:
not right now..